aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationProcessor.java

```diff
@@ -287,15 +296,24 @@ void stopDrainingMutationOutbox() {
         Class<T> modelClazz = (Class<T>) pendingMutation.getModelSchema().getModelClass();
         AppSyncConflictUnhandledError<T> unhandledConflict =
             AppSyncConflictUnhandledError.findFirst(modelClazz, errors);
-        if (unhandledConflict == null) {
-            return Single.error(new DataStoreException(
-                "Mutation failed. Failed mutation = " + pendingMutation + ". " +
-                    "AppSync response contained errors = " + errors,
-                "Verify that your AppSync endpoint is able to store " + modelClazz + " models."
-            ));
+        if (unhandledConflict != null) {
+            return conflictResolver.resolve(pendingMutation, unhandledConflict);
         }
 
-        return conflictResolver.resolve(pendingMutation, unhandledConflict);
+        // If error was not due to ConflictUnhandled, then just publish
+        // error to Hub and remove pending mutation from outbox.
+        // This helps unclog the mutation outbox by removing a failing
+        // mutation from the queue.
+        return Completable.fromAction(() -> announceMutationFailed(pendingMutation, errors))
+            .doOnError(error -> LOG.warn("Failed to create an outboxMutationFailedEvent. ", error))
+            .andThen(mutationOutbox.remove(pendingMutation.getMutationId()))
+            .andThen(Single.error(new DataStoreException(
+                    "Mutation failed. Failed mutation = " + pendingMutation + ". " +
+                            "AppSync response contained errors = " + errors,
+                    "Verify that your AppSync endpoint is able to store " +
+                            pendingMutation.getMutatedItem() + "."
+            )));
```

Comment:
Hmm.  I think this error is going to bubble up all the way to `MutationProcessor::startDrainingMutationOutbox`, and stop all subsequent mutations from being processed.  This is an existing problem, but it would be ideal if we can address it in this PR.

I would suggest these changes:
1. Instead of emitting a `DataStoreException`, emit a `DataStoreException.GraphQLResponseException`.  That will allow us to identify it as an error that should be swallowed higher up in the stack.
2. In `MutationProcessor::processOutboxItem`, if `publishToNetwork` gets an error, swallow it, and complete.  You can do that with something like this: 

```
.onErrorComplete(error -> error instanceof DataStoreException.GraphQLResponseException)
```

3. After making the first two changes, you could move the logic for the removing the item from the queue and calling `announceMutationFailed` into `processOutboxItem`.  Since `DataStoreException.GraphQLResponseException` is able to hold a list of `GraphQLResponse.Error` objects, they will be available to build a `OutboxMutationFailedEvent`.   It would be a little cleaner than doing it down in `handleResponseErrors`, since it would be right next to the call to `announceMutationProcessed`.  You'd add something like this:

```
.doOnError(error -> {
    mutationOutbox.remove(mutationOutboxItem.getMutationId())
        .doOnComplete(() -> announceMutationFailed(mutationOutboxItem, error))
}
```

Note that even with all these suggested changes, I think there is still a problem with the way non-AppSync errors are handled (e.g. network errors) (https://github.com/aws-amplify/amplify-android/issues/823).  We can address that separately though.

Reply:
Thanks for the detailed feedback! 
I will address the network errors in a separate PR.

Reply:
While doing some unrelated testing, I noticed that we have this existing problem around [line 120](https://github.com/aws-amplify/amplify-android/blob/8ccb280c0418853731f46fe709bda857d7b7765e/aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationProcessor.java#L120), too. If we fail to publish an item from the outbox (for example, if the app is offline), the exception tears down the observer, _and_ crashes the app. 😢